module/zfs/zil.c

```diff
@@ -2004,8 +2026,48 @@ zil_lwb_write_issue(zilog_t *zilog, lwb_t *lwb)
 	BP_ZERO(bp);
 	error = lwb->lwb_error;
 	if (error == 0) {
-		error = zio_alloc_zil(spa, zilog->zl_os, txg, bp, nlwb->lwb_sz,
-		    &slog);
+		/*
+		 * Allocation flexibility depends on LWB state:
+		 * if NEW: allow range allocation and larger sizes;
+		 * if OPENED: use fixed predetermined allocation size;
+		 * if CLOSED + Slim: allocate precisely for actual usage.
+		 */
+		boolean_t flexible = (nlwb->lwb_state == LWB_STATE_NEW);
+		if (flexible) {
+			/* We need to prevent opening till we update lwb_sz. */
+			mutex_enter(&nlwb->lwb_lock);
+			flexible = (nlwb->lwb_state == LWB_STATE_NEW);
+			if (!flexible)
+				mutex_exit(&nlwb->lwb_lock); /* We lost. */
+		}
+		boolean_t closed_slim = (nlwb->lwb_state == LWB_STATE_CLOSED &&
+		    nlwb->lwb_slim);
+
+		uint64_t min_size, max_size;
+		if (closed_slim) {
+			/* This transition is racy, but only one way. */
+			membar_consumer();
+			min_size = max_size = P2ROUNDUP_TYPED(nlwb->lwb_nused,
+			    ZIL_MIN_BLKSZ, uint64_t);
+		} else if (flexible) {
+			min_size = nlwb->lwb_min_sz;
+			max_size = nlwb->lwb_sz;
+		} else {
+			min_size = max_size = nlwb->lwb_sz;
+		}
+
+		error = zio_alloc_zil(spa, zilog->zl_os, txg, bp,
+		    min_size, max_size, &slog, flexible);
```

Comment:
Is there any danger calling this with `nlwb->lwb_lock` held? It'll shortly end up in `metaslab_alloc_range()` which will try to take `SCL_ALLOC`. I have no reason at all to suspect a problem, I'm just wary of those sort of things.

Reply:
The expected consequence is a blockage of `zil_lwb_write_open()` calls from other threads, that are also holding `zl_issuer_lock`.  So if we indefinitely stuck in allocation, we'll block any ZIL writes for this dataset.  I can only speculate some difference in context of your pool suspension recovery work, but I think it may be just between blocking one thread that is calling `zil_lwb_write_issue()` vs few more trying to fill the following LWB.  That is all assuming `metaslab_alloc_range()` doesn't have own mechanisms to recovery from pool suspension.